### PR TITLE
update filters tests and fix behaviours

### PR DESCRIPTION
**Pull request by suchen-sci**

### CODE REVIEW

File: pkg/filters/corsadaptor/corsadaptor.go [hunk 1]
```diff
@@ -120,24 +121,28 @@ func (a *CORSAdaptor) Handle(ctx *context.Context) string {
 
 func (a *CORSAdaptor) handle(ctx *context.Context) string {
 	r := ctx.Request().(*httpprot.Request)
-	w := ctx.Response().(*httpprot.Response)
 	method := r.Method()
 	headerAllowMethod := r.Header().Get("Access-Control-Request-Method")
 	if method == http.MethodOptions && headerAllowMethod != "" {
-		a.cors.HandlerFunc(w.Std(), r.Std())
+		rw := httptest.NewRecorder()
```
Comment (localvar) [inline pkg/filters/corsadaptor/corsadaptor.go:127]: I think this is ok, but would be better if we can avoid using `httptest`, as it it for testing.

File: pkg/filters/headertojson/headertojson.go [hunk 2]
```diff
@@ -123,6 +127,9 @@ func (h *HeaderToJSON) Handle(ctx *context.Context) string {
 		return ""
 	}
 
+	if fetcher, ok := ctx.Request().(fetchPayload); ok {
+		fetcher.FetchPayload()
+	}
```
Comment (localvar) [inline pkg/filters/headertojson/headertojson.go:132]: Don't need to do this.

File: pkg/filters/headertojson/headertojson.go [hunk 3]
```diff
@@ -64,6 +64,10 @@ type (
 		spec      *Spec
 		headerMap map[string]string
 	}
+
+	fetchPayload interface {
+		FetchPayload() (int, error)
+	}
```
Comment (localvar) [inline pkg/filters/headertojson/headertojson.go:70]: should be reverted, refer the next comment.

File: pkg/filters/kafkabackend/kafka.go [hunk 4]
```diff
@@ -59,6 +60,10 @@ type (
 		done     chan struct{}
 		header   string
 	}
+
+	fetchPayload interface {
+		FetchPayload() (int, error)
+	}
```
Comment (localvar) [inline pkg/filters/kafkabackend/kafka.go:66]: should revert, no need to call `FetchPayload` in most filters.

File: pkg/filters/remotefilter/remotefilter.go [hunk 5]
```diff
@@ -195,7 +195,8 @@ func (rf *RemoteFilter) limitRead(reader io.Reader, n int64) []byte {
 // Handle handles HTTPContext by calling remote service.
 func (rf *RemoteFilter) Handle(ctx *context.Context) (result string) {
 	r := ctx.Request().(*httpprot.Request)
-	w := ctx.Response().(*httpprot.Response)
+	w, _ := httpprot.NewResponse(nil)
+	ctx.SetResponse(ctx.TargetResponseID(), w)
```
Comment (localvar) [inline pkg/filters/remotefilter/remotefilter.go:199]: This is incorrect, should first check if the response already exists.

File: pkg/filters/requestadaptor/requestadaptor.go [hunk 6]
```diff
@@ -121,9 +122,29 @@ func (ra *RequestAdaptor) reload() {
...
+}
+
 // Handle adapts request.
 func (ra *RequestAdaptor) Handle(ctx *context.Context) string {
 	httpreq := ctx.Request().(*httpprot.Request)
+	if ra.spec.Body != "" || ra.spec.Compress != "" || ra.spec.Decompress != "" {
+		_, err := httpreq.FetchPayload()
+		if err != nil {
+			return resultReadBodyFail
+		}
```
Comment (localvar) [inline pkg/filters/requestadaptor/requestadaptor.go:145]: why call `FetchPayload`?

File: pkg/filters/validator/validator.go [hunk 7]
```diff
@@ -129,10 +129,20 @@ func (v *Validator) reload() {
...
+	httpresp, _ := httpprot.NewResponse(nil)
+	ctx.SetResponse(ctx.TargetResponseID(), httpresp)
+	return httpresp
+}
+
 // Handle validates HTTPContext.
 func (v *Validator) Handle(ctx *context.Context) string {
 	req := ctx.Request().(*httpprot.Request)
-	resp := ctx.Response().(*httpprot.Response)
+	resp := getResponse(ctx)
```
Comment (localvar) [inline pkg/filters/validator/validator.go:145]: should only create a response if there's an error.

Comment (suchen-sci) [reply pkg/filters/validator/validator.go]: updated!

**Review by localvar — APPROVED** [review]

**Pull request merged**